lark/grammars/lark.lark

```diff
@@ -7,46 +7,66 @@ _item: rule
      | token
      | statement
 
-rule: RULE rule_params priority? ":" expansions
-token: TOKEN token_params priority? ":" expansions
+rule: RULE_MODIFIERS? RULE rule_params priority? ":" rule_expansions
+token: TOKEN priority? ":" token_expansions
 
 rule_params: ["{" RULE ("," RULE)* "}"]
-token_params: ["{" TOKEN ("," TOKEN)* "}"]
 
 priority: "." NUMBER
 
-statement: "%ignore" expansions                    -> ignore
+statement: "%ignore" ignore_token                  -> ignore
          | "%import" import_path ["->" name]       -> import
          | "%import" import_path name_list         -> multi_import
          | "%override" rule                        -> override_rule
+         | "%override" token                       -> override_token
          | "%declare" name+                        -> declare
+         | "%extend" rule                          -> extend_rule
+         | "%extend" token                         -> extend_token
+
+ignore_token: ignore_item [ OP | "~" NUMBER [".." NUMBER]]
+ignore_item: STRING | TOKEN | REGEXP
 
 !import_path: "."? name ("." name)*
 name_list: "(" name ("," name)* ")"
 
-?expansions: alias (_VBAR alias)*
+?rule_expansions: rule_alias (_VBAR rule_alias)*
+
+?rule_inner_expansions: rule_expansion (_VBAR rule_expansion)*
+
+?rule_alias: rule_expansion ["->" RULE]
+
+?rule_expansion: rule_expr*
+
+?rule_expr: rule_atom [OP | "~" NUMBER [".." NUMBER]]
+?rule_atom: "(" rule_inner_expansions ")"
+          | "[" rule_inner_expansions "]" -> rule_maybe
+          | rule_value
+
+?rule_value: RULE "{" rule_value ("," rule_value)* "}" -> rule_template_usage
+           | RULE
+           | token_value
 
-?alias: expansion ["->" RULE]
+?token_expansions: token_expansion (_VBAR token_expansion)*
 
-?expansion: expr*
+?token_expansion: token_expr*
 
-?expr: atom [OP | "~" NUMBER [".." NUMBER]]
+?token_expr: token_atom [OP | "~" NUMBER [".." NUMBER]]
 
-?atom: "(" expansions ")"
-     | "[" expansions "]" -> maybe
-     | value
+?token_atom: "(" token_expansions ")"
```

Comment:
token_atom and rule_atom are the same, why not merge them using templates? Same with rule_expr and atom_expr, rule_expansion etc.

Reply:
Before we dig into that, let's come to an agreement in the main conversation (regarding restrictive-validating grammar vs relaxed grammar + validating visitor).

Afterwards if it's still relevant, I will try to craft a solution to this problem.